tutorials/realm/MachineModel.md

```diff
@@ -0,0 +1,134 @@
+* [Processor](#processor)
+* [Memory](#memory)
+* [Affinity](#affinity)
+* [ID](#id)
+* [References](#references)
+
+## Machine
+
+In Realm, a Machine is the highest level of machine model, representing all the computer nodes an application can occupy. 
+It can be retrieved using the `Machine::get_machine()` function, 
+which returns a singleton object.
+Processors, memories of the `Machine`, and affinity information can be queried from the `Machine` object.
+
+## Processor
+
+A `Processor` is a runtime object that represents any execution resource that can run a task, such as CPU, GPU, etc.
+To retrieve a Processor, `ProcessorQuery` can be used. In this example, we use `ProcessorQuery` to iterate over all 
+enabled processors and print out their address space and ID (a unique identifier Realm assignes).
+```c++
+for(Machine::ProcessorQuery::iterator it = Machine::ProcessorQuery(machine).begin(); it; ++it) {
+  ...
+}
+```
+The address space of a Realm resource, such as Processor and Memory (introduced in the next section), indicates the process/rank where it resides.
+
+A `Processor` can also be queried by its `kind` and the runtime supports the following options:
+
+- `TOC_PROC` represents a throughput processor, which is usually a CPU core. 
+It can be specified by `-ll:cpu`.
+- `LOC_PROC` represents a latency processor (GPU).Currently, Realm supports both NVIDIA and AMD GPUs.
+It can be specified by `-ll:gpu`.
+- `UTIL_PROC` represents a CPU processor that is designed for users to run their own background work.
+It can be specified by `-ll:util`.
+- `IO_PROC` represents a processor that is used for I/O, which is also a CPU core. 
+It can be specified by `-ll:io`.
+- `PROC_GROUP` represents a group of processors. 
+- `PROC_SET` represents a set of processors for OpenMP/Kokkos etc. It can be specified by `-ll:mp_nodes`.
+- `OMP_PROC` represents OpenMP thread pool. It can be specified by `-ll:ocpu`. The number of threads per `OMP_PROC` can be specified by `-ll:othr`.
+- `PY_PROC` represents a CPU processor that is used for Python interpreter. 
+It can be specified by `-ll:py`. Currently, we only support a single `PY_PROC`.
+
+For a list of all the processors kind supported by Realm, please refer to [Full Processor Kind](#full-proc-kind).
+
+## Memory
+
+`Memory` is used to describe the location of application data. `MemoryQuery` can be used to query 
+the Memory. For example, a `MemoryQuery` can be created with the condition `has_affinity_to` to return 
+all memories that are affixed to the given processor.
+```c++
+Machine::MemoryQuery mq = Machine::MemoryQuery(machine).has_affinity_to(p, 0, 0);
+```
+
+A `Memory` can also be queried by its `kind` and the runtime supports the following options:
+
+- `GLOBAL_MEM` represents CPU memory guaranteed to be visible to all processors on all nodes.
+e.g. GASNet global memory. `GLOBAL_MEM` is usually slow. `GLOBAL_MEM` is only used by MPI and GASNet1 modules, 
+and it can be specified by `-ll:gsize`.
```

Comment:
When you create list items, please indent the text to 4 spaces. Again, this is for compatibility with other Markdown parsers, which may not all parse this the same way.

```
  - First line of next.
    Next line of text.
    Last line of text.
  - First line of second bullet.
    ...
```

Reply:
fixed for all realm tutorials.